sdk/core/Azure.Core/src/Pipeline/HttpPipeline.cs

```diff
@@ -124,7 +129,68 @@ public Response SendRequest(Request request, CancellationToken cancellationToken
         /// </example>
         public static IDisposable CreateClientRequestIdScope(string? clientRequestId)
         {
-            return ReadClientRequestIdPolicy.StartScope(clientRequestId);
+            return CreateHttpMessagePropertiesScope(new Dictionary<string, object?>() { { ReadClientRequestIdPolicy.MessagePropertyKey, clientRequestId } });
+        }
+
+        /// <summary>
+        /// Creates a scope in which all <see cref="HttpMessage"/>s would have provided properties.
+        /// </summary>
+        /// <param name="messageProperties">Properties to be added to <see cref="HttpMessage"/>s</param>
+        /// <returns>The <see cref="IDisposable"/> instance that needs to be disposed when properties shouldn't be used anymore.</returns>
+        public static IDisposable CreateHttpMessagePropertiesScope(IDictionary<string, object?> messageProperties)
+        {
+            Argument.AssertNotNull(messageProperties, nameof(messageProperties));
+            CurrentHttpMessagePropertiesScope.Value = new HttpMessagePropertiesScope(messageProperties, CurrentHttpMessagePropertiesScope.Value);
+            return CurrentHttpMessagePropertiesScope.Value;
+        }
+
+        private static void AddHttpMessageProperties(HttpMessage message)
+        {
+            if (CurrentHttpMessagePropertiesScope.Value != null)
+            {
+                foreach (KeyValuePair<string, object?> kvp in CurrentHttpMessagePropertiesScope.Value.Properties)
+                {
+                    if (kvp.Value != null)
```

Comment:
Does it throw if you don't check this?  We can wait and see, but I wonder if there are valid customer scenarios for clearing a value already on the message.

Reply:
The [HttpMessage._properties] doesn't like nulls. But we figured that it's good to have a way to support scenarios like this (already existing for request-id scope) https://github.com/Azure/azure-sdk-for-net/blob/4fb4680de87735063f5e1c0209206c72d934a321/sdk/core/Azure.Core/src/HttpMessage.cs#L17 .
I.e. that user can "unset" some property from parent scope, without affecting parent.